// @medusajs/medusa-js/dist/resources/admin/tax-rates.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminSwapsResource = /** @class */ (function (_super) {
    __extends(AdminSwapsResource, _super);
    function AdminSwapsResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminSwapsResource.prototype.retrieve = function (id, query) {
        var path = "/admin/tax-rates/" + id;
        return this.client.request("GET", path);
    };
    AdminSwapsResource.prototype.list = function (query) {
        var path = "/admin/tax-rates";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/tax-rates?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    AdminSwapsResource.prototype.create = function (payload, query) {
        var path = "/admin/tax-rates";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/tax-rates?" + queryString.join("&");
        }
        return this.client.request("POST", path, payload);
    };
    AdminSwapsResource.prototype.update = function (id, payload, query) {
        var path = "/admin/tax-rates/" + id;
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/tax-rates/" + id + "?" + queryString.join("&");
        }
        return this.client.request("POST", path, payload);
    };
    AdminSwapsResource.prototype.addProducts = function (id, payload, query) {
        var path = "/admin/tax-rates/" + id + "/products";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/tax-rates/" + id + "/products?" + queryString.join("&");
        }
        return this.client.request("POST", path, payload);
    };
    AdminSwapsResource.prototype.addProductTypes = function (id, payload, query) {
        var path = "/admin/tax-rates/" + id + "/product-types";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/tax-rates/" + id + "/product-types?" + queryString.join("&");
        }
        return this.client.request("POST", path, payload);
    };
    AdminSwapsResource.prototype.addShippingOptions = function (id, payload, query) {
        var path = "/admin/tax-rates/" + id + "/shipping-options";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/tax-rates/" + id + "/shipping-options?" + queryString.join("&");
        }
        return this.client.request("POST", path, payload);
    };
    AdminSwapsResource.prototype.delete = function (id) {
        var path = "/admin/tax-rates/" + id;
        return this.client.request("DELETE", path);
    };
    return AdminSwapsResource;
}(base_1.default));
exports.default = AdminSwapsResource;
